---
title: "Data science with R for Life scientist course project"
author: "Glory Ifeoluwa Jayeoba"
date: "08/09/2025"
output:
  pdf_document: default
  html_notebook: default
  html_document:
    df_print: paged
---


**Project 1: RNA differential gene expression analysis**

This R notebook is a documentation of the step by step process of analyzing a biological data as part of the HackBio data science with R for Life scientist beginners course. The dataset used in this project is a pre-processed RNAseq data from an experiment between a healthy cell line and diseased cell lines treated with compound X. The difference in expression change between the two health status is computed as Fold change to log 2 (Log2FChange) and the significance of each is computed in p-value. 



*Step 1: Setting up the R environment*

To begin analysis, you will need to set up your R environment and that include ensuring that all the packages needed as available. To install R packages needed if not available in the R library, use the install.package() function. 

For this analysis, the packages required aside from the functions available in the R base environment are:
ggplot2 
tidyverse
RColorBrewer
ggrepel
```{r}
#loading all packages needed in the R environment
library(ggplot2)        ##similar to the plot() function. 
library(tidyverse)      ##collection of packages for data import and manipulation
library(RColorBrewer)   
library(ggrepel)        ##for managing overlapping text in plot labels
```

*Step 2: Loading dataset from Local file.*

Once the R environment is set, the next thing to do is to upload your dataset into your R. Uploading datasets will depend on the location of the file. You can assess the file from web links, your local file in PC etc. The process of importing a file in R is called reading files. Generally, there are several types of file depending on the delimiter e.g comma separated files, tab separated files etc. For each of these type of file, there are corresponding functions in R used to read them. 
For example, if the file is a csv (comma separated file), you can use the read.csv (). Another way is to use the read.table() function and specify the delimiter of the file using the "sep = ". 
It is important to note that if the delimiter is not specified or the wrong delimiter is specified, R is read all the column as a single column.

There are other packages in R that has read table functions. For instance, readr is a part of the tidyverse package that performs similar function as read.table. read_delim (which will be used in this project)is a part of the readr package that reads file separated by space.

```{r}
#loading the data and assigning it to an object named dataset.
getwd()
dataset <- read_delim("C:/Users/HP/OneDrive/Documents/dataset_DumpDeseq.txt", 
                      delim = " ", 
                      col_names = TRUE)
dataset
```

Note the data was referred to as a tibble which is the same as a dataframe.
The uploaded file has 4 columns and 16406 rows. The first column contains a list of hit genes identified from the RNAseq experiment. The second column contains the log2 of the foldchange which is the difference in gene expression change between the two health status. The third column contains the pvalue that is the significance of these changes. The last column contains the adjusted p-values to eliminate possible false positives.



*Step 3: Cleaning up data*

Depending on the state of your data, you may be required to perform several clean up. This may include converting data to the appropriate class, filtering out missing data etc.
the %>% is the syntax for pipe. This means the result from the first function will be used for the next function

```{r}
#checking for missing values in all the columns in the data
dataset <- dataset %>% filter(!is.na(log2FoldChange), 
                              !is.na(pvalue), 
                              !is.na(padj))
dataset
```
After filtering, the number of columns and rows remained the same implying that there is no missing data in the dataset.


*Step 4: Creating a basic volcano plot*

To create the volcano plot, the log2foldchange was plotted against the -log10 of the pvalues. Pvalues are relatively small numbers and if used directly, most points will be below 0 making it difficult to interpret the plot. Using the -log10(pvalue) will convert the pvalues to larger positive values making it easier to observe which value is most significant. 

In this project the package used for the plot is ggplot2. However, the plot function the R base can be used as well.

```{r}
#using ggplot to create the volcano plot
ggplot(data = dataset, aes(x = log2FoldChange, #setting the values x and y axis.
                           y = -log10(pvalue))) + 
  geom_point() 

#using the plot()function to create the volcano plot
plot(dataset$log2FoldChange,        #setting the values for the x and y axis
     -log10(dataset$pvalue),    
     xlab = "log2FoldChange",       #setting the label for the x axis
     ylab = "-log10(pvalue)",       #setting the label for the y axis                              
     col = "black",                 #setting the points color to black
     pch = 19)                      #setting the shape of the points 
```

This will produce a basic volcano plot for the data.

*Step 5:Identifying up-regulated and down-regulated genes*

Since the log2foldchange values shows the difference between the healthy and diseased state, a value 0 will mean no difference, a positive value will mean a positive difference (up-regulated genes) while a negative value will mean negative difference (down regulated gene).
By setting a threshold, we can identify genes that are up-regulated, down-regulated and genes with no difference.

```{r}
##Setting threshold
## Up-regulated gene = Log2FoldChange > 1 and pvalue < 0.01
## Down-regulated gene = Log2FoldChange < -1 and pvalue < 0.01

#using ggplot package
ggplot(data = dataset, aes(x = log2FoldChange,
                           y = -log10(pvalue))) +    #setting the x and y axis
  geom_vline(xintercept = c(-1, 1), 
             col = "gray", 
             linetype = "dashed") +      #setting the threshold for the x axis
  geom_hline(yintercept = -log10(0.01),
             col = "gray", 
             linetype = "dashed") +      #setting the threshold for y axis 
  geom_point()


#using the plot() function
plot(dataset$log2FoldChange, -log10(dataset$pvalue),  #setting the values for axes
     xlab = "log2FoldChange",                         #setting the x label
     ylab = "-log10(pvalue)",                   #setting the y label
     col = "black",                             #setting the color to black
     pch = 19,                                  #setting the shape of points
     bg = "black")                          #setting the background of the points
abline(h = -log10(0.01), 
       col = "gray", 
       lty = 2)                 #horizontal threshold to have gray dashed lines
abline(v = c(-1, 1), 
       col = "gray", 
       lty = 2)                 #vertical threshold to have gray dashed lines

```

All values at the left side of the vertical line indicate down regulated genes, while those on the right side of the vertical line represents up-regulated genes. All values below the horizontal lines represent genes that are not significant while those above the line represent genes that are significant. 

We can also color each gene based on their expression levels to indicate those that are up-regulated and the down-regulated. To achieve this, first we need to create  anew column categorizing each gene based on their expression level. All values in the expression column was first set to "NO", then the up or down regulated genes were assigned either "UP" or "DOWN" 
```{r}
##creating a new column that categorize each gene.
dataset$expression <- "NO"          #setting all values in the new column to NO

#setting values within this threshold to UP
dataset$expression[dataset$log2FoldChange >1 & dataset$pvalue < 0.01] <- "UP" 

#setting values within this threshold to DOWN
dataset$expression[dataset$log2FoldChange < -1 & dataset$pvalue < 0.01] <- "DOWN" 

#using ggplot package
ggplot(data = dataset, 
       aes(x = log2FoldChange, y = -log10(pvalue), col = expression)) + 
  geom_vline(xintercept = c(-1, 1),
             col = "gray", 
             linetype = "dashed") + 
  geom_hline(yintercept = -log10(0.01), 
             col = "gray", 
             linetype = "dashed") + 
  geom_point() +
  
  #setting the colors, blue(down), grey(no), red(up)
  scale_color_manual(values = c("blue", "grey", "red"),                  
                     labels = c("downregulated", "not significant",
                                "upregulated"))

##Using the plot() function
#assign colors based on expression values (similar to scale_color_manual)
colour <- c("DOWN" = "blue", "NO" = "gray", "UP" = "red")

plot(dataset$log2FoldChange, -log10(dataset$pvalue),
     xlab = "log2FoldChange",
     ylab = "-log10(pvalue)", 
     bg = "gray",
     pch = 19,
     col = colour[dataset$expression])  
#adding legend to the plot
legend("bottomright", 
       legend = c("downregulated", "not significant", "upregulated"),   
       levels(as.factor(dataset$expression)), 
       col = colour,
       pch = 19,
       cex = 0.8,
       bty = "n")
```

Finally, we can annotate the plot by adding gene labels. To do that, we need to create a new column that only names the 5 top down and up regulated genes while others are set to NA. We are creating a new column instead of using the gene symbol column because if we use the gene symbol column, all the points on the plot will be labeled with their gene name. However, we are only interested in the top 5 genes. Think of creating this new column as a means of filtering out our genes of interest and setting the rest to NA 

```{r}
#create a new column that contains the top 5 up and down regulated genes.

#Selecting the top 5 up-regulated gene
top_up <- dataset %>% 
  filter(log2FoldChange > 1) %>%
  arrange(pvalue) %>%
  slice_head(n = 5) %>%
  pull(Gene)

#this means: For values in dataset, filter those with log2FoldChange > 1, 
#arrange the result based on their pvalues (lowest to highest), 
#extract the top 5 rows, and extract all the values in the Gene column

#Selecting the top 5 down-regulated gene (similar to top_up)
top_down <- dataset %>%
  filter(log2FoldChange < -1) %>%
  arrange(pvalue) %>%
  slice_head(n = 5) %>%
  pull(Gene)

#Setting all other gene aside the top up and down regulated genes to NA
dataset$topgenes <- ifelse(dataset$Gene %in% c(top_up, top_down),
                           dataset$Gene, NA)

#plotting with the top 5 gene labeled
#Using ggplot package
ggplot(data = dataset, 
       aes(x = log2FoldChange, y = -log10(pvalue), col = expression)) + 
  geom_point() +
  scale_color_manual(values = c("blue", "grey", "red"), 
                     labels = c("downregulated", "not significant",
                                "upregulated")) +
  geom_text_repel(aes(label = topgenes), max.overlaps = 15) + 
  theme_minimal() +
  theme_classic()



#Using plot () function
colour <- c("DOWN" = "blue", "NO" = "gray", "UP" = "red")
plot(dataset$log2FoldChange, -log10(dataset$pvalue),
     xlab = "log2FoldChange",
     ylab = "-log10(pvalue)", 
     bg = "gray",
     pch = 19,
     col = colour[dataset$expression])
legend("bottomright", 
       legend = c("downregulated", "not significant", "upregulated"),
       levels(as.factor(dataset$expression)), 
       col = colour,
       pch = 19,
       cex = 0.8,
       bty = "n")
topdata <- subset(dataset, !is.na(topgenes))
with(topdata, text(
  log2FoldChange, -log10(pvalue), 
  labels = topgenes,
  col = ifelse(log2FoldChange > 0, "red", "blue"),
  cex = 0.7,
  pos = ifelse(log2FoldChange > 0, 2, 2)
))

```
Adding final features such as title, re-scaling, etc.
```{r}
#for plot using ggplot package
ggplot(data = dataset, 
       aes(x = log2FoldChange, y = -log10(pvalue), col = expression)) + 
  geom_point() +
  scale_color_manual(values = c("blue", "grey", "red"), 
                     labels = c("downregulated", "not significant",
                                "upregulated")) +
  geom_text_repel(aes(label = topgenes), max.overlaps = 15) + 
  theme_minimal() +
  theme_classic() +
  ggtitle("Differential expression of genes between healthy and diseased cell lines response to compund X") +
  scale_x_continuous(limits = c(-3, 3), breaks = seq(-3,3,1)) +
  scale_y_continuous(limits = c(0, 10), breaks = seq(0,10,2)) +
```


```{r}
theme(plot.title = element_text(hjust = 1, size = 8, face = "bold"))



#For plot using plot () function
par(bty = "l")
colour <- c("DOWN" = "blue", "NO" = "gray", "UP" = "red")
plot(dataset$log2FoldChange, -log10(dataset$pvalue),
     xlab = "log2FoldChange",
     ylab = "-log10(pvalue)", 
     xlim = c(-3, 3),
     ylim = c(0, 10),
     bg = "gray",
     pch = 19,
     cex = 0.8,
     title(main = "Differential expression of genes between healthy and diseased cell lines response to compund X",
           cex.main = 0.7,
           adj = 0.5),
     col = colour[dataset$expression])
legend("topright", 
       legend = c("Expression", "Downregulated", "Not significant", "Upregulated"),
       levels(as.factor(dataset$expression)), 
       col = c(NA, colour),
       pch = c(NA, 19, 19, 19),
       cex = 0.8,
       bty = "n",
       text.font = c(2,1,1,1))
axis(1, at = seq(-3, 3, 1))
axis(2, at = seq(0, 10, 2))
topdata <- subset(dataset, !is.na(topgenes))
with(topdata, text(
  log2FoldChange, -log10(pvalue), 
  labels = topgenes,
  col = ifelse(log2FoldChange > 0, "red", "blue"),
  cex = 0.7,
  pos = ifelse(log2FoldChange > 0, 2, 2)
))
```

**The top 5 up-regulated genes are:**

EMILIN2: May be responsible for anchoring smooth muscle cells to elastic fibers, and may be involved not only in the formation of the elastic fiber, but also in the processes that regulate vessel assembly.
Has cell adhesive capacity. 

POU3F4: This gene encodes a member of the POU-III class of neural transcription factors. This family member plays a role in inner ear development. The protein is thought to be involved in the mediation of epigenetic signals which induce striatal neuron-precursor differentiation. Mutations in this gene are associated with X chromosome-linked nonsyndromic mixed deafness.

LOC285954: 
VEPH1: Predicted to enable phosphatidylinositol-5-phosphate binding activity. Involved in negative regulation of SMAD protein signal transduction and negative regulation of transforming growth factor beta receptor signaling pathway

DTHD1: this gene encodes a protein which contains a death domain. Death domain-containing proteins function in signaling pathways and formation of signaling complexes, as well as the apoptosis pathway. Alternative splicing results in multiple transcript variants.  

**The top 5 down regulated genes are:**

TBX5: This gene is a member of a phylogenetically conserved family of genes that share a common DNA-binding domain, the T-box. T-box genes encode transcription factors involved in the regulation of developmental processes. This gene is closely linked to related family member T-box 3 (ulnar mammary syndrome) on human chromosome 12. The encoded protein may play a role in heart development and specification of limb identity. Mutations in this gene have been associated with Holt-Oram syndrome, a developmental disorder affecting the heart and upper limbs.

LAMA2: Laminin, an extracellular protein, is a major component of the basement membrane. It is thought to mediate the attachment, migration, and organization of cells into tissues during embryonic development by interacting with other extracellular matrix components. It is composed of three subunits, alpha, beta, and gamma, which are bound to each other by disulfide bonds into a cross-shaped molecule. This gene encodes the alpha 2 chain, which constitutes one of the subunits of laminin 2 (merosin) and laminin 4 (s-merosin). Mutations in this gene have been identified as the cause of congenital merosin-deficient muscular dystrophy.

CAV2:The protein encoded by this gene is a major component of the inner surface of caveolae, small invaginations of the plasma membrane, and is involved in essential cellular functions, including signal transduction, lipid metabolism, cellular growth control and apoptosis. This protein may function as a tumor suppressor. This gene and related family member (CAV1) are located next to each other on chromosome 7, and express colocalizing proteins that form a stable hetero-oligomeric complex. 

IFITM1: Interferon-induced transmembrane (IFITM) proteins are a family of interferon induced antiviral proteins. The family contains five members, including IFITM1, IFITM2 and IFITM3 that belong to the CD225 superfamily. The protein encoded by this gene restricts cellular entry by diverse viral pathogens, such as influenza A virus, Ebola virus and Sars-CoV-2. 

TNN: Predicted to enable integrin binding activity. Involved in positive regulation of sprouting angiogenesis; regulation of cell adhesion; and regulation of cell migration. Part of tenascin complex

